File: zul/src/main/java/org/zkoss/zul/Chip.java
/* Chip.java

	Purpose:

	Description:

	History:
		Wed Apr 22 2026, Created by yuehfeng

Copyright (C) 2026 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
package org.zkoss.zul;

import org.zkoss.lang.Objects;
import org.zkoss.zk.ui.WrongValueException;
import org.zkoss.zk.ui.event.Events;
import org.zkoss.zk.ui.ext.Disable;
import org.zkoss.zul.impl.LabelImageElement;
import org.zkoss.zul.impl.Utils;

/**
 * A display-only chip used to show categories, statuses, or keywords,
 * with an optional close button and a severity theme.
 *
 * <p><b>There is no {@code color} attribute on this component.</b> To
 * customize appearance, either set the {@code style} attribute (e.g.
 * {@code style="--zk-chip-bg: #6366f1;"}) or override the CSS custom
 * properties {@code --zk-chip-bg} / {@code --zk-chip-color} /
 * {@code --zk-chip-border} declared in {@code chip.less}.
 *
 * <p>Default {@link #getZclass}: "z-chip".
 *
 * @author yuehfeng
 * @since 10.4.0
 */
public class Chip extends LabelImageElement implements Disable {
	private static final long serialVersionUID = -7935159928926565441L;
	static {
		addClientEvent(Chip.class, Events.ON_CLOSE, 0);
	}

	private static final String DEFAULT_SEVERITY = "info";
	private static final String DEFAULT_SIZE = "medium";

	private String _severity = DEFAULT_SEVERITY;
	private String _size = DEFAULT_SIZE;
	private boolean _rounded;
	private boolean _closable;
	private boolean _disabled;

	public Chip() {
	}

	public Chip(String label) {
		super(label);
	}

	/** Returns the severity (color theme).
	 * <p>Default: "info".
	 */
	public String getSeverity() {
		return _severity;
	}

	/** Sets the severity.
	 * @param severity one of "info", "success", "warning", "danger", "secondary".
	 *        Pass null to reset to the default ("info").
	 */
	public void setSeverity(String severity) throws WrongValueException {
		severity = Utils.checkEnum(severity, DEFAULT_SEVERITY,
				"severity must be info/success/warning/danger/secondary: ",
				"info", "success", "warning", "danger", "secondary");
		if (!Objects.equals(_severity, severity)) {
			_severity = severity;
			smartUpdate("severity", _severity);
		}
	}

	/** Returns the size of this chip.
	 * <p>Default: "medium". One of "small" or "medium"; medium maps to the
	 * default 22px row, small reduces the row height for dense layouts.
	 */
	public String getSize() {
		return _size;
	}

	/** Sets the chip size.
	 * @param size "small" or "medium". Pass null to reset to the default ("medium").
	 */
	public void setSize(String size) throws WrongValueException {
		size = Utils.checkEnum(size, DEFAULT_SIZE, "size must be small or medium: ", "small", "medium");
		if (!Objects.equals(_size, size)) {
			_size = size;
			smartUpdate("size", _size);
		}
	}

	/** Whether this chip is non-interactive. Disabled chips ignore close
	 * clicks and render at reduced opacity.
	 */
	public boolean isDisabled() {
		return _disabled;
	}

	/** Sets whether this chip is non-interactive. Disabled chips ignore close
	 * clicks and render at reduced opacity.
	 * @param disabled whether the chip is disabled. Calls smartUpdate when changed.
	 */
	public void setDisabled(boolean disabled) {
		if (_disabled != disabled) {
			_disabled = disabled;
			smartUpdate("disabled", _disabled);
		}
	}

	/** Whether this chip is rounded (pill-shaped). */
	public boolean isRounded() {
		return _rounded;
	}

	/** Sets whether this chip is rounded (pill-shaped).
	 * @param rounded whether the chip is rounded. Calls smartUpdate when changed.
	 */
	public void setRounded(boolean rounded) {
		if (_rounded != rounded) {
			_rounded = rounded;
			smartUpdate("rounded", _rounded);
		}
	}

	/** Whether this chip displays a close button. */
	public boolean isClosable() {
		return _closable;
	}

	/** Sets whether this chip displays a close button.
	 * @param closable whether the chip displays a close button. Calls smartUpdate when changed.
	 */
	public void setClosable(boolean closable) {
		if (_closable != closable) {
			_closable = closable;
			smartUpdate("closable", _closable);
		}
	}

	//super//
	@Override
	public String getZclass() {
		return _zclass == null ? "z-chip" : _zclass;
	}

	@Override
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer) throws java.io.IOException {
		super.renderProperties(renderer);
		if (!"info".equals(_severity))
			render(renderer, "severity", _severity);
		if (!"medium".equals(_size))
			render(renderer, "size", _size);
		if (_rounded)
			render(renderer, "rounded", _rounded);
		if (_closable)
			render(renderer, "closable", _closable);
		if (_disabled)
			render(renderer, "disabled", _disabled);
	}

	/** Handles the onClose event sent when the close button is clicked.
	 * <p>Default: detach this chip. Override or use {@code event.stopPropagation()}
	 * to keep the chip visible. Follows the same pattern as
	 * {@link Window#onClose} — the framework's event-dispatch machinery
	 * skips the default handler when a prior listener stops propagation.
	 * @since 10.4.0
	 */
	public void onClose() {
		// The close button is rendered disabled client-side, but a
		// hand-crafted onClose AU request could still reach here; the server
		// must not trust the client's disabled enforcement.
		if (isDisabled())
			return;
		detach();
	}

	@Override
	public boolean isChildable() {
		return false;
	}
}

File: zul/src/main/resources/web/js/zul/wgt/Chip.ts
/* Chip.ts

        Purpose:
                
        Description:
                
        History:
                Mon May 11 15:04:46 CST 2026, Created by peakerlee

Copyright (C) 2026 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/

/**
 * A display-only chip.
 * @defaultValue {@link getZclass}: "z-chip".
 */
@zk.WrapClass('zul.wgt.Chip')
export class Chip extends zul.LabelImageWidget {
	/** @internal */ _severity = 'info';
	/** @internal */ _size = 'medium';
	/** @internal */ _rounded = false;
	/** @internal */ _closable = false;
	/** @internal */ _disabled = false;

	/**
	 * Returns the severity (color theme).
	 * @defaultValue `info`.
	 */
	getSeverity(): string { return this._severity; }
	/**
	 * Sets the severity (color theme). null restores the default (`info`).
	 * @param severity - one of "info", "success", "warning", "danger" or
	 * "secondary".
	 */
	setSeverity(severity: string, opts?: Record<string, boolean>): this {
		// Mirror Avatar/Java null→default coercion: a null/undefined from a
		// stateless MVVM binding would otherwise reach $s() and emit a dead
		// `z-chip-null` class with no LESS rule.
		if (severity == null) severity = 'info';
		const o = this._severity;
		this._severity = severity;
		if (o !== severity || opts?.force) this.updateDomClass_();
		return this;
	}

	/**
	 * Returns the size of this chip. "medium" maps to the default row height;
	 * "small" reduces it for dense layouts.
	 * @defaultValue `medium`.
	 */
	getSize(): string { return this._size; }
	/**
	 * Sets the size of this chip. null restores the default (`medium`).
	 * @param size - "small" or "medium".
	 */
	setSize(size: string, opts?: Record<string, boolean>): this {
		if (size == null) size = 'medium';
		const o = this._size;
		this._size = size;
		if (o !== size || opts?.force) this.updateDomClass_();
		return this;
	}

	/**
	 * Returns whether this chip is non-interactive. Disabled chips ignore close
	 * clicks and render at reduced opacity.
	 * @defaultValue `false`.
	 */
	isDisabled(): boolean { return this._disabled; }
	/**
	 * Sets whether this chip is non-interactive. Disabled chips ignore close
	 * clicks and render at reduced opacity.
	 * @param disabled - whether the chip is disabled.
	 */
	setDisabled(disabled: boolean, opts?: Record<string, boolean>): this {
		const o = this._disabled;
		this._disabled = disabled;
		if (o !== disabled || opts?.force) {
			this.updateDomClass_();
			// The mold renders <button disabled> at HTML time based on
			// _disabled, but a runtime setDisabled(true) only flips the
			// CSS class — leaving the close button still keyboard-/click-
			// activatable, which surprises a screen-reader user navigating
			// a "disabled" chip. Sync the native attribute too.
			var closeBtn = this.$n('close');
			if (closeBtn) {
				if (disabled) closeBtn.setAttribute('disabled', 'disabled');
				else closeBtn.removeAttribute('disabled');
			}
		}
		return this;
	}

	/**
	 * Returns whether this chip is rounded (pill-shaped).
	 * @defaultValue `false`.
	 */
	isRounded(): boolean { return this._rounded; }
	/**
	 * Sets whether this chip is rounded (pill-shaped).
	 * @param rounded - whether the chip is rounded.
	 */
	setRounded(rounded: boolean, opts?: Record<string, boolean>): this {
		const o = this._rounded;
		this._rounded = rounded;
		if (o !== rounded || opts?.force) this.updateDomClass_();
		return this;
	}

	/**
	 * Returns whether this chip displays a close button.
	 * @defaultValue `false`.
	 */
	isClosable(): boolean { return this._closable; }
	/**
	 * Sets whether this chip displays a close button.
	 * @param closable - whether the chip displays a close button.
	 */
	setClosable(closable: boolean, opts?: Record<string, boolean>): this {
		const o = this._closable;
		this._closable = closable;
		if (o !== closable || opts?.force) this.rerender();
		return this;
	}

	/**
	 * @returns the widget's CSS class.
	 * @defaultValue `z-chip`.
	 */
	override getZclass(): string {
		return this._zclass == null ? 'z-chip' : this._zclass;
	}

	/** @internal */
	override domClass_(no?: zk.DomClassOptions): string {
		var /*safe*/ sc = super.domClass_(no);
		if (!no || !no.zclass) {
			sc += ' ' + this.$s(this._severity);
			if (this._size !== 'medium') sc += ' ' + this.$s(this._size);
			if (this._rounded) sc += ' ' + this.$s('rounded');
			if (this._closable) sc += ' ' + this.$s('closable');
			if (this._disabled) sc += ' ' + this.$s('disabled');
		}
		return sc;
	}

	/** @internal */
	override doClick_(evt: zk.Event): void {
		if (this._disabled) {
			evt.stop();
			return;
		}
		if (this._closable) {
			var closeNode = this.$n('close'),
				target = evt.domTarget as HTMLElement | undefined;
			// Match the close button itself OR anything inside it (the
			// icon-font glyph child <i> receives the click in real browsers).
			if (closeNode && target && (target === closeNode || closeNode.contains(target))) {
				this._fireClose();
				evt.stop();
				return;
			}
		}
		super.doClick_(evt);
	}

	/** @internal */
	_fireClose(): void {
		// Let the server decide visibility: the default onClose handler
		// detaches the chip (the client gets an invalidate response), but a
		// listener may call event.stopPropagation() to suppress that detach
		// — see Chip.java onClose(). Optimistically hiding here would leave
		// the chip invisible but still attached in the suppressed case.
		this.fire('onClose', {}, {toServer: true});
	}
}
